Моисеев Н.И. ИУ3-33Б ЛР2

60Hukuma92 opened the pull request.

Vlaikivanikz committed:
init

Vlaikivanikz committed:
Лабораторная работа 1

Vlaikivanikz committed:
added comments

Vlaikivanikz committed:
ошибки исправлены

Vlaikivanikz committed:
array new method

IU3Labs committed:
Merge pull request #11 from Vlaikivanikz/ivanovvr-1

Лабораторная работа 1

IU3Labs committed:
Revert "Лабораторная работа 1"

IU3Labs committed:
Merge pull request #33 from IU3Labs/revert-11-ivanovvr-1

Revert "Лабораторная работа 1"

60Hukuma92 committed:
Main task completed

60Hukuma92 committed:
Group A task completed

60Hukuma92 committed:
Group C task RBT class is completed. Still need to write test for it though.

60Hukuma92 committed:
Group B task LRUCache completed. All test for the collections completed.
Lab 2 is considered finished.

IU3Labs commented on lab/lab2/groupA/CircularQueue.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public void enqueue(T element) {
+        if (isFull()) throw new IllegalStateException("CircularQueue is overloaded");
```
Цикличность не отражена

60Hukuma92 replied:
Возможно, вы имеете в виду недостаток комментариев по логике работы цикличной очереди, добавил их.

**CircularQueue** - СД, основанная на массиве. Её цикличность позволяет записывать элементы в конец очереди фактически в начало массива, если там освободилось место, то есть некоторые элементы были удалены из очереди. При этом **цикличность** - это только **про реализацию очереди**, а **не про функционал**. Он совпадает с простым Queue. Прикрепляю иллюстрирующую картинку.
<img width="860" height="412" alt="ircularQueuePicture" src="https://github.com/user-attachments/assets/0a1cc51e-2cc2-43e1-a032-9cd8c447dcf8" />

Циклическое (круговое) перемещение индекса по формуле                           _i_ = (_i_ + 1) % _CAPACITY_
Обеспечивает корректный переход через границу массива - гарантирует, что индекс всегда остается в границах [0, _CAPACITY_-1]. 
%CAPACITY при перемещении указателя rear обеспечивает то, что после достижения индекса _CAPACITY_-1, в тот момент, когда i должен стать равным _CAPACITY_, остаток от деления перемещает указатель на индекс 0, ведь _CAPACITY_ % _CAPACITY_ даёт 0. Всем индексам, меньшим _CAPACITY_, а именно такими они и являются, %_CAPACITY_ не мешает, ведь он даёт именно эти же индексы.

Если под цикличностью подразумевается перезаписывание элементов, то это уже напоминает такую СД, как **Cache**, ведь из очереди элементы не должны пропадать (перезаписываться).

IU3Labs replied:
Не зачтено

60Hukuma92 committed:
Comments about CircularQueue added. Some other comments were corrected as well.

60Hukuma92 commented:
Постарался объяснить цикличность CircularQueue, если я правильно вас понял.

IU3Labs commented:
Решить конфликт

60Hukuma92 committed:
Update CircularQueue.java

CircularQueue is now overwriting its elements when size == CAPACITY. Now it does not throw IlligalStateException when it is full.

60Hukuma92 committed:
Trying to resolve the conflict

60Hukuma92 committed:
Still trying to resolve the conflict

60Hukuma92 commented:
Всё сделал как надо
